dockerpty/io.py

```diff
@@ -121,23 +124,61 @@ def read(self, n=4096):
 
     def write(self, data):
         """
-        Write `data` to the Stream.
+        Write `data` to the Stream. Not all data may be written right away.
+        Use select to find when the stream is writeable, and call do_write()
+        to flush the internal buffer.
         """
 
         if not data:
             return None
 
+        self.buffer += data
+        self.do_write()
+
+        return len(data)
+
+    def do_write(self):
+        """
+        Flushes as much pending data from the internal write buffer as possible.
+        """
         while True:
             try:
+                written = 0
+
                 if hasattr(self.fd, 'send'):
-                    self.fd.send(data)
-                    return len(data)
-                os.write(self.fd.fileno(), data)
-                return len(data)
+                    written = self.fd.send(self.buffer)
+                else:
+                    written = os.write(self.fd.fileno(), self.buffer)
```

Comment:
I'd like to abstract this difference behind some interface, but I don't think it's blocking (since we were doing this before as well)